src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -193,39 +200,59 @@ public override void OnWriterCompleted(Action<Exception, object> callback, objec
         }
 
         /// <inheritdoc />
-        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        public override ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
         {
             // TODO ReadyAsync needs to throw if there are overlapping reads.
             ThrowIfCompleted();
 
-            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
-            CancellationTokenSource tokenSource = InternalTokenSource;
-            if (TryReadInternal(tokenSource, out ReadResult readResult))
+            if (TryReadInternal(out ReadResult readResult))
             {
-                return readResult;
+                return new ValueTask<ReadResult>(readResult);
             }
 
             if (_isStreamCompleted)
             {
-                return new ReadResult(buffer: default, isCanceled: false, isCompleted: true);
+                return new ValueTask<ReadResult>(new ReadResult(buffer: default, isCanceled: false, isCompleted: true));
+            }
+
+            if (_streamReadTask.IsCompleted)
+            {
+                _streamReadTask = ReadStreamAsync(cancellationToken);
+
+                // Completed the stream read inline because it was synchronous and there was no exception thrown
+                if (_streamReadTask.IsCompleted && _edi == null)
+                {
+                    return new ValueTask<ReadResult>(GetReadResult());
+                }
             }
 
-            var reg = new CancellationTokenRegistration();
+            return new ValueTask<ReadResult>(this, 0);
```

Comment:
Might as well add a `short` field and count up prior to this call e.g.
```csharp
_token++;
return new ValueTask<ReadResult>(this, _token);
```
Then check if `token` passed to `GetResult(short token)` is the same and throw if not?

Reply:
We don’t check the token today in other the default pipe, I’m not going to add it here unless we do it in both. I’d store then token as part of the PipeAwaitable state